libz-rs-sys-cdylib/src/gz.rs

```diff
@@ -2571,6 +2571,93 @@ pub unsafe extern "C-unwind" fn gzrewind(file: gzFile) -> c_int {
     0
 }
 
+/// # Safety
+///
+/// The variadic arguments must correspond with the format string in number and type.
+#[cfg(feature = "gzprintf")]
+#[export_name = crate::prefix!(gzprintf)]
+pub unsafe extern "C-unwind" fn gzprintf(
+    file: gzFile,
+    format: *const c_char,
+    mut va: ...
+) -> c_int {
+    unsafe { gzvprintf(file, format, va.as_va_list()) }
+}
+
+#[cfg(feature = "gzprintf")]
+unsafe extern "C-unwind" fn gzvprintf(
+    file: gzFile,
+    format: *const c_char,
+    va: core::ffi::VaList,
+) -> c_int {
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return Z_STREAM_ERROR;
+    };
+
+    // Check that we're writing and that there's no error.
+    if state.mode != GzMode::GZ_WRITE || state.err != Z_OK {
+        return Z_STREAM_ERROR;
+    }
+
+    // Make sure we have some buffer space.
+    if state.input.is_null() && gz_init(state).is_err() {
+        return state.err;
+    }
+
+    // Check for seek request.
+    if state.seek {
+        state.seek = false;
+        if gz_zero(state, state.skip as _).is_err() {
+            return state.err;
+        }
+    }
+
+    // Do the printf() into the input buffer, put length in len -- the input
+    // buffer is double-sized just for this function, so there is guaranteed to
+    // be state.size bytes available after the current contents
+    if state.stream.avail_in == 0 {
+        state.stream.next_in = state.input;
+    }
+
+    extern "C" {
+        fn vsnprintf(
+            s: *mut c_char,
+            n: libc::size_t,
+            format: *const c_char,
+            va: core::ffi::VaList,
+        ) -> i32;
+    }
+
+    let next = state.input.wrapping_add(
+        (unsafe { state.stream.next_in.offset_from(state.input) })
+            .wrapping_add(state.stream.avail_in as isize) as usize,
+    );
+    let state_size = state.in_size / 2;
```

Comment:
This fixes an out-of-bounds read. This makes some sense to me intuitively, but I'm not sure it is correct. We should really document how `in_size` and `out_size` relate to the `size` used by zlib-ng. Can you provide some details there?

Reply:
The printf implementation in zlib-ng depends on this:
```
    /* do the printf() into the input buffer, put length in len -- the input
       buffer is double-sized just for this function, so there is guaranteed to
       be state->size bytes available after the current contents */
```

which suggests that zlib-ng's other write functions only ever half-fill the `in` buffer. That's currently not the case for zlib-rs, where the other gz-related write functions currently allow the `input` buffer to fill up completely before flushing it.

Reply:
I'd prefer to follow `zlib-ng` here. I don't think performance is really a concern.

Reply:
how hard would it be to instead have `in_size` be the same value as `zlib-ng`? (and allocate twice that number of bytes in read mode, similarly for write)

Reply:
#378